### PR TITLE
feat(glue): Connection construct

### LINKED ISSUE #12442: (aws-glue): add Connection construct
In https://github.com/aws/aws-cdk/issues/5172 and its PRs, the suggestion was made to break a PR down into 2
1. Add Connection construct
2. Add Job construct

I've since had a look and realized the work needed involves 3 constructs
1. Connection construct (this request)
2. SecurityConfiguration construct #12449
3. Job construct #12443

Since the original issue and PR have gone without progress for a time, I am contributing some effort here and as such will be creating 3 issues and hopefully 3 PRs to address each of the issues.

This is the first issue; to add the Glue `Connection` construct

### Use Case

Be able to create Glue jobs with L2 constructs that utilize Glue connections. My immediate use case is to be able to connect data sources in a VPC

### Proposed Solution

Introduce a `Connection` L2 construct wrapping the [Glue Connection CloudFormation resource](https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-glue-connection.html)

### Others
Some relevant links
- https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-resource-glue-connection.html
- https://docs.aws.amazon.com/glue/latest/dg/connection-using.html
- https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-connections.html#aws-glue-api-catalog-connections-Connection

* [x] :wave: I may be able to implement this feature request
* [ ] :warning: This feature might incur a breaking change

---

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by humanzz**
Closes #12442

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by humanzz** [comment]
I will add integration tests as well so this is likely not final.

File: packages/@aws-cdk/aws-glue/lib/connection.ts [hunk 1]
```diff
@@ -0,0 +1,193 @@
...
+
+/**
+ * The type of the glue connection
+ *
+ * If you need to use a connection type that doesn't exist as a static member, you
+ * can instantiate a `ConnectionType` object, e.g: `new ConnectionType('NEW_TYPE')`.
+ */
+export class ConnectionType {
+  /** A list of all known `ConnectionType`s. */
+  public static readonly ALL = new Array<ConnectionType>();
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/lib/connection.ts:14]: Why is this needed?

File: packages/@aws-cdk/aws-glue/lib/connection.ts [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+
+  /**
+   * Designates a connection to a database through Java Database Connectivity (JDBC).
+   */
+  public static readonly JDBC = new ConnectionType('JDBC');
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/lib/connection.ts:19]: Why not enums? Are these likely to have new values frequently?

File: packages/@aws-cdk/aws-glue/lib/connection.ts [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+   *  Key-Value pairs that define parameters for the connection.
+   *  @default empty properties
+   */
+  readonly connectionProperties?: { [key: string]: string };
+
+  /**
+   * A list of criteria that can be used in selecting this connection.
+   * @default no match criteria
+   */
+  readonly matchCriteria?: string[];
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/lib/connection.ts:112]: An example and/or documentation of these would be good.

File: packages/@aws-cdk/aws-glue/lib/connection.ts [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+export class Connection extends cdk.Resource implements IConnection {
+
+  /**
+   * Creates a Connection construct that represents an external connection.
+   *
+   * @param scope The scope creating construct (usually `this`).
+   * @param id The construct's id.
+   * @param attrs Import attributes
+   */
+  public static fromConnectionAttributes(scope: constructs.Construct, id: string, attrs: ConnectionAttributes): IConnection {
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/lib/connection.ts:139]: split to `fromConnectionName` and `fromConnectionArn`. We usually reserve `fromXXAttributes` to allow importing additional runtime information, i.e those used beyond the ID of the resource.

File: packages/@aws-cdk/aws-glue/lib/connection.ts [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+  /**
+   * The description of the connection.
+   * @default no description
+   */
+  readonly description?: string;
+
+  /**
+   * The type of the connection
+   */
+  readonly connectionType: ConnectionType;
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/lib/connection.ts:100]:
```suggestion
  readonly type: ConnectionType;
```

File: packages/@aws-cdk/aws-glue/lib/connection.ts [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+
+  /**
+   *  Key-Value pairs that define parameters for the connection.
+   *  @default empty properties
+   */
+  readonly connectionProperties?: { [key: string]: string };
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/lib/connection.ts:106]:
```suggestion
  readonly properties?: { [key: string]: string };
```

File: packages/@aws-cdk/aws-glue/lib/connection.ts (hunk 1 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts:14]: Answering this comment and the following one about `ConnectionType`
- A few months back, `JDBC` was the only supported type
- When I started working on this PR, I did notice the new supported types
- I basically followed the pattern I noticed in https://github.com/aws/aws-cdk/blob/master/packages/%40aws-cdk/aws-lambda/lib/runtime.ts#L39

Having said all of that, I'm happy to either
1. Switch to a normal enum (am not sure when is the next time they'll change or how frequent it is) or
1. Make all private

What do you think?

File: packages/@aws-cdk/aws-glue/lib/connection.ts (hunk 2 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts:19]: See my response above

File: packages/@aws-cdk/aws-glue/lib/connection.ts (hunk 4 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts:139]: will do that

File: packages/@aws-cdk/aws-glue/lib/connection.ts (hunk 5 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts:100]: will do that

File: packages/@aws-cdk/aws-glue/lib/connection.ts (hunk 6 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts:106]: will do that

File: packages/@aws-cdk/aws-glue/lib/connection.ts (hunk 3 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts:112]: Added a further comment on where this can be used

File: packages/@aws-cdk/aws-glue/lib/connection.ts (hunk 1 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts]: I looked further - the glue cloudformation documentation is not super great on this area to be honest - and found mentions of even more types in https://docs.aws.amazon.com/glue/latest/dg/aws-glue-programming-etl-connect.html

This makes me a bit inclined to keeping it extensible rather than an enum

**Comment by humanzz** [comment]
@iliapolo I think I might have wrongly rebased things which brought in unnecessary commits into this diff. Apologies for that

File: packages/@aws-cdk/aws-glue/lib/connection.ts (hunk 1 above)
Comment (iliapolo) [reply packages/@aws-cdk/aws-glue/lib/connection.ts]: I agree the documentation is not great. and I also agree it seems to be a more open system than typical enums. The documentation you referred to made me consider a different approach though. 

Since every connection type comes with its own connection options, it would be great to provide compile time assistance for that. Basically, we provide type-safe connection types, as well as raw connections as a sort of escape hatch:

```ts
// type safe API. type specification is not necessary, and all options are modeled and enforced according to
// https://docs.aws.amazon.com/glue/latest/dg/aws-glue-programming-etl-connect.html#aws-glue-programming-etl-connect-kafka
const kafkaConnection = new glue.KafkaConnection(this, 'KafkaConnection', {
  bootstrapServers: ['http://server1.kafka'],
  topicName: 'TheTopic',
  ...
});

// even type-safe API's will have an escape hatch:
kafkaConnection.addOption('key', 'value');

// escape hatch, free for all.
const parquetConnection = new glue.Connection(this, 'ParquetConnection', {
  type: 'parquet', // string as is, no enums here. 
  options: { // any key value 
  }
})
```

Would you be up to doing 2-3 of these type safe connections? I know it might be a little tedious :), but once we have that pattern locked-down, it will be easy to incrementally add type-safe API's by the community.

WDYT?

Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts]: Wait till you see the `Job` construct because I keep feeling it does require way more compile-time assistance.
I shied away from doing that for the size of work required to be honest and I kept thinking can we do something now that we can build on later without breaking the APIs. Let's continue that `Job` discussion for when we get there.

For this `Connection` work though, let me have a think about it. Maybe we separate it into ensuring we have the escape hatch version first and separate the work to add more compile-type safe versions on top.

I'll have a look and see what I can do and report back

Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts]: So, I have a couple of thoughts/questions
The main changes to support your suggestion would be to
1. Introduce new concrete classes with new props that are more geared towards the connection types
2. Be able to map their options into the proper options names
3. Provide that `addOption` or `addProperty` escape hatch to be able to add more things

In my mind, this can be achieved basically by extending the current `Connection`... maybe extract some `BaseConnectionProps` that drops `properties` in favour of each connection's options.

What I am not clear about - a pattern I haven't coded before - is the escape hatch... it basically means that the base resource cannot be created right away in the constructor, right? are there relevant methods/hooks to use to ensure that the base resource gets instantiated/synthesized after all escape hatch calls `addOption` are done?

Comment (iliapolo) [reply packages/@aws-cdk/aws-glue/lib/connection.ts]: @humanzz 

> What I am not clear about - a pattern I haven't coded before - is the escape hatch... it basically means that the base resource cannot be created right away in the constructor, right? are there relevant methods/hooks to use to ensure that the base resource gets instantiated/synthesized after all escape hatch calls `addOption` are done?

The way we do this is by using `Lazy` values. It basically means that instead of assigning a concrete value to the underlying CFN property, you assign a function, that gets evaluated only at synth time. 

For example, you instantiate the underlying resource in the constructor as usual:

https://github.com/aws/aws-cdk/blob/9c5000dc3e0a750f43ffa620b3bef60c694387fc/packages/%40aws-cdk/aws-rds/lib/parameter-group.ts#L139-L143

Notice the `parameters` property is assigned with a `Lazy.any` that accepts a function, delaying the access to `this.parameters`.

This way - post instantiation additions are supported as well:

https://github.com/aws/aws-cdk/blob/9c5000dc3e0a750f43ffa620b3bef60c694387fc/packages/%40aws-cdk/aws-rds/lib/parameter-group.ts#L170-L173

For the sake of speediness though, I'm also perfectly ok with skipping the `addOption` method for now and just providing a `custom` key in the constructor:

```ts
const kafkaConnection = new glue.KafkaConnection(this, 'KafkaConnection', {
  topicName: 'MyTopic',
  custom: {
    'key': 'value',
  }
})
```

Whatever you feel you have capacity for.

Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/connection.ts]: Nice... So I made some changes to the `Connection` and its props. I've also added an example `MongoDBConnection` class.
I don't feel comfortable adding other types coz testing that they actually do work and are passing the correct params would be quite time consuming because one has to really fill in the documentation gaps and test and iterate.

So, how about we go with what we have, if we think the pattern is correct and allows non-breaking future extensibility, and leave adding more type-safe implementations for whoever will be needing them and having more capacity to contribute?

File: packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+   * - MongoPaginateByCountPartitioner
+   * - MongoPaginateBySizePartitioner
+   *
+   * @default MongoDefaultPartitioner
+   */
+  readonly partitioner?: string;
+}
+
+/**
+ * An AWS Glue connection to a MongoDB source.
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts:57]: Add a link to the docs.

File: packages/@aws-cdk/aws-glue/test/connection.test.ts [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+    resourceName: connectionName,
+  }));
+});
+
+test('fromConnectionArn', () => {
+  const connectionArn = 'arn:aws:glue:region:account-id:connection/name';
+  const stack = new cdk.Stack();
+  const connection = glue.Connection.fromConnectionArn(stack, 'ImportedConnection', connectionArn);
+
+  strictEqual(connection.connectionName, 'name');
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/test/connection.test.ts:158]: use `expect`

File: packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts [hunk 9]
```diff
@@ -0,0 +1,83 @@
+import * as constructs from 'constructs';
+import { Connection, ConnectionOptions, ConnectionType } from './connection';
+
+/**
+ * Construction properties for {@link MongoDBConnection}
+ *
+ * @see https://docs.aws.amazon.com/glue/latest/dg/connection-mongodb.html
```
[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts:7]: Renamed to source connection. The differences am noticing are around `partitionerOptions` which I avoided adding coz it itself is a bunch of extra properties and documentation was not clear about how to best name them.

[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts:7]: @iliapolo fixed merge conflicts and now this is ready for another round of review... can't wait till we get to the `Job` one, it will be... mmm... something :-)

[the remark this answers is not in the document]
Comment (iliapolo) [reply packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts:7]: What about `uri` that is mentioned as a required option in the below documentation link? And `username` and `password`?

File: packages/@aws-cdk/aws-glue/lib/connection.ts [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+
+  /**
+   * The description of the connection.
+   * @default no description
+   */
+  readonly description?: string;
+
+  /**
+   *  Key-Value pairs that define parameters for the connection.
+   *  @default empty properties
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/lib/connection.ts:86]:
```suggestion
   *  @default empty properties
   *  @see https://docs.aws.amazon.com/glue/latest/dg/aws-glue-programming-etl-connect.html
```

File: packages/@aws-cdk/aws-glue/README.md [hunk 11]
```diff
@@ -23,6 +23,55 @@
...
+new glue.Connection(stack, 'MyConnection', {
+  connectionType: glue.ConnectionTypes.NETWORK,
+  securityGroups: [securityGroup],
+  subnet,
+});
+```
+
+If you need to use a connection type that doesn't exist as a static member on `ConnectionType`, you can instantiate a `ConnectionType` object, e.g: `new glue.ConnectionType('NEW_TYPE')`.
+
+Different connection types require different connection properties which can be set via `ConnectionProps.properties`.
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/README.md:40]: No need.

File: packages/@aws-cdk/aws-glue/README.md [hunk 12]
```diff
@@ -23,6 +23,55 @@
...
+  securityGroups: [securityGroup],
+  subnet,
+});
+```
+
+If you need to use a connection type that doesn't exist as a static member on `ConnectionType`, you can instantiate a `ConnectionType` object, e.g: `new glue.ConnectionType('NEW_TYPE')`.
+
+Different connection types require different connection properties which can be set via `ConnectionProps.properties`. 
+
+Subclasses of `Connection` are implemented to provide a connection type tailored props e.g. `MongoDBSourceConnection` which can be used as follows
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/README.md:42]:
```suggestion
Concrete connection types are also available and provide a tailored, type-safe API for the specific connection options. 

### MongoDB
```

File: packages/@aws-cdk/aws-glue/README.md [hunk 13]
```diff
@@ -23,6 +23,55 @@
...
+
+```ts
+new glue.MongoDBSourceConnection(stack, 'Connection', {
+    database: 'database',
+    collection: 'collection',
+});
+```
+
+if a specific connection type construct props does not expose all properties, additional properties can still be set as follows
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/README.md:51]:
```suggestion
In case a specific connection type construct does not expose all properties, additional properties can still be set using the `properties` property.
```

File: packages/@aws-cdk/aws-glue/README.md (hunk 11 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/README.md:40]: Do you mean line 40 or 38 as well? I think for 38 you mentioned it should be here - from the doc class docs?

File: packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts (hunk 9 above)
[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts:7]: You are right, I missed those. However, I'm still not really sure about the right thing here
- https://docs.aws.amazon.com/glue/latest/dg/connection-mongodb.html: The connection `url`, `username` and `password` are stored in the MongoDB connection. Other options can be specified in the job script as additional options.
- https://docs.aws.amazon.com/glue/latest/dg/connection-mongodb.html: `uri` instead of `url`

What I did try is, using the console to create a mongodb connection. It prompted me for connection url which itself includes a databasename, username, password, and vpc configs.

I think I need to add an integ test to try a couple of variations to figure out the right properties to use to match what gets created in console... I can't say it enough, documentation is really horrible!

[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts:7]: Neither `url` or `uri` seem to work. I've also checked https://docs.aws.amazon.com/glue/latest/dg/aws-glue-api-catalog-connections.html#aws-glue-api-catalog-connections-ConnectionInput which lists a different set of potential values for properties keys.

What do you think about dropping the whole MongoDB connection in favour of the now ready-to-extend `Connection`? And maybe reserve those type-safe classes for when the documentation is clearer?

[the remark this answers is not in the document]
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/lib/mongodb-connection.ts]: After further discussions, we'll drop the MongoDB connection construct while keeping the escape hatches to ensure the idea can be still implemented once we have clearer requirements/documentation on each connection type requirements.

File: packages/@aws-cdk/aws-glue/README.md (hunk 11 above)
Comment (humanzz) [reply packages/@aws-cdk/aws-glue/README.md]: Dropped L40

File: packages/@aws-cdk/aws-glue/README.md [hunk 14]
```diff
@@ -23,6 +23,22 @@
...
 
+## Connection
+
+A `Connection` allows Glue jobs, crawlers and development endpoints to access certain types of data stores.
+
+```ts
+new glue.Connection(stack, 'MyConnection', {
+  connectionType: glue.ConnectionTypes.NETWORK,
+  securityGroups: [securityGroup],
+  subnet,
```
Comment (iliapolo) [inline packages/@aws-cdk/aws-glue/README.md:34]: If we have in the example, lets just add a sentence or two into what these are - and why they are needed.

Comment (humanzz) [reply packages/@aws-cdk/aws-glue/README.md]: Basing my comments on Subnet/Security groups section in https://docs.amazonaws.cn/en_us/glue/latest/dg/connection-defining.html

**Review by iliapolo — APPROVED** [review]

**Review by iliapolo — APPROVED** [review]
